### PR TITLE
Parse INI (for Android)

### PR DESCRIPTION
**Pull request by jahorton**
Parallels #451.

The aim is for this class to parallel the iOS version of the PR as well as possible, meeting the same specifications, etc.

Further detail on our use of INI and that spec is necessary to proceed further.

### CODE REVIEW

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * A custom, lightweight INI parser designed initially for use with Keyman's .kmp and .inf files.
+ * @author jahorton
+ *
+ */
+public class INIParser {
```
Comment (jahorton) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java]: Fun detail - Android Studio requires any desired test classes and tests to be declared `public`.  This differs from the situations I've experienced in other IDEs.

If a test class itself is not public, Android Studio won't realize that it exists as a testing class.  Caused me quite the headache for a while.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
+package com.tavultesoft.kmea.util;
+
+import java.util.Map;
+import java.util.Scanner;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * A custom, lightweight INI parser designed initially for use with Keyman's .kmp and .inf files.
```
Comment (darcywong00) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java]: We should probably reference the kmp.inf spec. At the moment, it lives in a Google doc.
Maybe we should persist that on the keymanapp/keyman wiki where it's publicly visible?

Comment (mcdurdin) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java]: Gabe put it online at https://github.com/keymanapp/keyman/wiki/KMP-Metadata-File-(kmp.inf). That's now the master reference.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ * @author jahorton
+ *
+ */
+public class INIParser {
+	public static Map<String, Map<String, List<String>>> parse(String _string) throws INIParseException {
+		Map<String, Map<String, List<String>>> sections = new HashMap<String, Map<String, List<String>>>();
+		
+		String[] linesArray = _string.split("\n");
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java]: Will this cope with `\r\n` from Windows?

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+		
+		// If the default section is empty, remove it.
+		if(sections.get("").isEmpty()) {
+			sections.remove("");
+		}
+		
+		return sections;
+	}
+	
+	public static List<String> components (String text) throws INIParseException {
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java]: I think we should avoid doing this in the `INIParser` class: technically the INI format is just name=value pairs. This way we can use the class on arbitrary INI files in the future if we had reason to.

Also some values in the .inf format may have "," which are part of their value, and not separating components, for example `[Keyboard0].Name` may have a value with a comma in it. Thus, this componentization should probably only be done for the entries that specifically require it, according to the kmp.inf spec

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java (hunk 2 above)
Comment (jahorton) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java]: Probably my biggest question/concern:  what are legal characters within individual elements of value lists?  What about within double-quoted elements?

File: android/KMEA/app/src/test/java/com/tavultesoft/kmea/util/INIParserTests.java [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+import org.junit.Assert;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.junit.Test;
+
+public class INIParserTests {
```
Comment (jahorton) [inline android/KMEA/app/src/test/java/com/tavultesoft/kmea/util/INIParserTests.java]: Fun detail - Android Studio requires any desired test classes and test functions to be declared `public`.  This differs from the situations I've experienced in other IDEs.

If a test class itself is not public, Android Studio won't realize that it exists as a testing class.  Caused me quite the headache for a while.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+		} finally {
+			tokens.close();
+		}
+		
+		return values;
+	}
+	
+	static String unquote(String text) throws INIParseException {
+		// Filter both types of quotes.
+		String val = unquote_helper(text, '\'', '\'',
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java]: I didn't realise this but you are correct; single quotes are also stripped by `GetPrivateProfileString` for values. Wow. Need to make sure this gets into the iOS PR as well.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	}
+	
+	static String unquote_helper(String text, char open, char closed, String unclosedMsg, String malformedMsg) throws INIParseException {
+		if(text.charAt(0) == open) {
+			text = text.substring(1);
+			int closeIndex = text.indexOf(closed);
+			
+			if(closeIndex == -1) {
+				throw new INIParseException(unclosedMsg + ": " + text);
+			} else if(closeIndex != text.length() - 1) {
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java]: Instead, of looking for the first closing quote mark, we need to look at the last character in the (trimmed) string. For example `"a"b"c""` is considered valid by `GetPrivateProfileString` and will be returned as `a"b"c"`. Unbalanced quotes return the full string, and don't throw. We only throw at a higher level when parsing componentized strings such as in `[Files].i`

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java (hunk 3 above)
Comment (jahorton) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java]: Split will automatically leave a `\r` on the end of each line.  Fortunately, that's technically a whitespace character, meaning `trim()` will nix it.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java (hunk 4 above)
Comment (jahorton) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java]: > Thus, this componentization should probably only be done for the entries that specifically require it, according to the kmp.inf spec.

Got it, that'll be a part two thing.  Though... that leaves me wondering about this situation.

```
value="foo","bar"
value2="foobar"
```

How do we ensure we don't, well, FUBAR either value there?

**Comment by mcdurdin** [comment]
Per discussion in Standup, we're abandoning .inf format and moving to .xml for mobile devices. And the world said hurrah.

**Pull request closed without merging**